blog-posts/typescript-ast-type-checker.md

```diff
@@ -0,0 +1,591 @@
+      if (ts.isTypeAliasDeclaration(node) && node.name.escapedText === "Output") {
+        const outputType = checker.getTypeAtLocation(node.name);
+        const [collectedValueProperty] = outputType.getProperties();
+
+        /**
+         * `propertyType` will contain & reference everything
+         * we can know about the type `ImportantValue`
+         */
+        const propertyType = checker.getTypeOfSymbolAtLocation(collectedValueProperty, node);
+        const propertyTypeName = checker.typeToString(propertyType);
+        // prints `collectedValue: ImportantValue`
+        console.log(`── ${collectedValueProperty.name}: ${propertyTypeName}`)
+
+        for (const nestedProperty of propertyType.getProperties()) {
+          const nestedPropertyType = checker.getTypeOfSymbolAtLocation(nestedProperty, node);
+          const nestedPropertyTypeName = checker.typeToString(nestedPropertyType);
+          /** prints the following
+           ├── value1: string
+           ├── value2: number
+           ├── value3: Date
+           */
+          console.log(`     ├── ${nestedProperty.name}: ${nestedPropertyTypeName}`)
+        }
+      }
+    });
+    ````
+</details>
+
+## Real-world adjustments
+
+The basic demonstration was specifically crafted to demonstrate the type extraction process, but there are some important real-world issues when doing so.
+
+- We don't know the depth of our analysis, so it's a perfect match for recursion although you could construct a loop too I guess.
+- We need to prevent diving into properties that are coming from the standard library like `Date` and methods or values of primitives like `string` because we are usually not interested in those properties. Same for external libraries (think of rxjs & friends).
+
+### Recursion
+
+First, let's make the analysis recursive to find every property in any given file.
+
+```typescript
+function processProperty(type: ts.Type, node: ts.Node, level = 0) {
+  if(level === 0) {
+    console.group(`.\n└──Processing '${checker.typeToString(type)}'`)
+  }
+
+  for (const property of type.getProperties()) {
+    const propertyType = checker.getTypeOfSymbolAtLocation(property, node);
+    const propertyTypeName = checker.typeToString(propertyType);
+
+    processProperty(propertyType, node, level + 1)
+    console.log(`  ├── ${property.name}: ${propertyTypeName}`)
+
+  }
+  console.groupEnd();
+
+}
+
+ts.forEachChild(myComponentSourceFile, node => {
+  if (ts.isTypeAliasDeclaration(node) && node.name.escapedText === "Output") {
+    const outputType = checker.getTypeAtLocation(node.name);
+    processProperty(outputType, node);
+  }
+});
+```
+
+This will find every single property, no matter how deep it's nested. That's because `processProperty()` is used recursively on all nested properties.
+
+When you run this code, you will be lost in noise. See the following log and try to spot our types within the ocean of properties pouring in from the standard library.
+
+<details>
```

Comment:
`margin-bottom` is missing here. We could fix this separately.

<img width="909" alt="Screenshot 2021-06-21 at 08 02 46" src="https://user-images.githubusercontent.com/7814926/122714183-33f39d00-d267-11eb-9931-06d56c7c3182.png">

Reply:
I added 16px but we still need to check how this behaves in detail.